### PR TITLE
Fan export

### PR DESCRIPTION
**Pull request by pascutto**
This is built on top of #20.
It writes the fan out when merging so that read only and restarting instances can read this (small) file instead of the whole index.

### CODE REVIEW

File: src/index.ml [hunk 1]
```diff
@@ -112,6 +112,57 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
...
+
+    external get_64 : string -> int -> int64 = "%caml_string_get64"
+
+    external swap64 : int64 -> int64 = "%bswap_int64"
+
+    let encode_int64 i =
+      let set_uint64 s off v =
+        if not Sys.big_endian then set_64 s off (swap64 v) else set_64 s off v
+      in
+      let b = Bytes.create 8 in
```
Comment (samoht) [inline src/index.ml]: you can allocate that buffer only once outside of the function. As you do a copy (with `Bytes.to_string`) that should be safe (unless that code is used in threads...)

File: src/index.ml [hunk 2]
```diff
@@ -426,7 +468,13 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
       |> EntrySet.elements
     in
     merge_with log t tmp;
+    let tmp_fan_path = t.root // "tmp" // "fan" in
+    let fan_path = t.root // "fan" in
+    let tmp_fan = IO.v ~fresh:true ~readonly:false ~generation tmp_fan_path in
+    Fan.export t.fan_out tmp_fan;
+    let fan = IO.v ~fresh:true ~readonly:false ~generation fan_path in
     IO.rename ~src:tmp ~dst:t.index;
+    IO.rename ~src:tmp_fan ~dst:fan;
```
Comment (samoht) [inline src/index.ml]: beware: this is not atomic anymore. What happens it the program crashes between line 476 and 477?

**Comment by samoht** [comment]
If possible, I think it would be better if we keep all of these information into the same file, e.g. put the fanout at the beginning of the file.

**Comment by pascutto** [comment]
I don't think we can do this with the current append-only IO, can we ?

File: src/index.ml (hunk 1 above)
Comment (samoht) [reply src/index.ml]: (actually this is a very bad idea in the context of multi-threaded programs :p)

**Comment by samoht** [comment]
Can't you serialise the fan-out first and then add the entries? Or keep some space to fill it at the end?

**Comment by pascutto** [comment]
No satisfying solution for now. Closing.

**Pull request closed without merging**